contracts/src/v0.8/shared/call/CallWithExactGasZKSync.sol

```diff
@@ -0,0 +1,121 @@
+  /// of the `_maxTotalGas` parameter. In other words, `max(gas(), _maxTotalGas)` is the maximum amount of gas that can be spent by this function.
+  /// @dev The function relays the `returndata` returned by the callee. In case the `callee` reverts, it reverts with the same error.
+  /// @param _to The address of the contract to call.
+  /// @param _maxtotalgas the maximum amount of gas that can be spent by the call.
+  /// @param _data The calldata for the call.
+  /// @param _maxReturnBytes the maximum amount of bytes that can be returned by the call.
+  /// @return success whether the call succeeded
+  /// @return retData the return data from the call, capped at maxReturnBytes bytes
+  /// @return gasUsed the gas used by the external call. Does not include the overhead of this function.
+  function _callWithExactGasSafeReturnData(
+    address _to,
+    uint256 _maxTotalGas,
+    bytes calldata _data,
+    uint16 _maxReturnBytes
+  ) external payable {
+    // At the start of the execution we deduce how much gas be spent on things that will be
+    // paid for later on by the transaction.
+    // The `expectedForCompute` variable is an upper bound of how much this contract can spend on compute and
+    // MUST be higher or equal to the `gas` passed into the call.
+    uint256 expectedForCompute = gasleft() + CALL_ENTRY_OVERHEAD;
+
+    // We expect that the `_maxTotalGas` at least includes the `gas` required for the call.
+    // This require is more of a safety protection for the users that call this function with incorrect parameters.
+    //
+    // Ultimately, the entire `gas` sent to this call can be spent on compute regardless of the `_maxTotalGas` parameter.
+    require(_maxTotalGas >= gasleft(), "Gas limit is too low");
+
+    // This is the amount of gas that can be spent *exclusively* on pubdata in addition to the `gas` provided to this function.
+    uint256 pubdataAllowance = _maxTotalGas > expectedForCompute ? _maxTotalGas - expectedForCompute : 0;
+
+    uint256 pubdataPublishedBefore = SYSTEM_CONTEXT_CONTRACT.getCurrentPubdataSpent();
+
+    // We never permit system contract calls.
+    // If the call fails, the `EfficientCall.call` will propagate the revert.
+    // Since the revert is propagated, the pubdata published wouldn't change and so no
+    // other checks are needed.
+    bytes memory returnData = EfficientCall.call({
+      _gas: gasleft(),
+      _address: _to,
+      _value: msg.value,
+      _data: _data,
+      _isSystem: false
+    });
+
+    // We will calculate the length of the returndata to be used at the end of the function.
+    // We need additional `96` bytes to encode the offset `0x40` for the entire pubdata,
+    // the gas spent on pubdata as well as the length of the original returndata.
+    // Note, that it has to be padded to the 32 bytes to adhere to proper ABI encoding.
+    uint256 paddedReturndataLen = returnData.length + 96;
+    if (paddedReturndataLen % 32 != 0) {
```

Comment:
Is this needed? We don't do this in all the other callWithExactGas implemenations

Reply:
Not sure since this part I ported over from zksync's implementation here: https://github.com/matter-labs/era-contracts/blob/main/gas-bound-caller/contracts/GasBoundCaller.sol#L72

Reply:
We can probably improve a bit on their solution, as theirs assumes they copy all the calldata

Reply:
Removed, thanks!